builtin/aws/ecs/task.go

```diff
@@ -0,0 +1,249 @@
+package ecs
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/waypoint/builtin/aws/utils"
+	"github.com/oklog/ulid"
+
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	"github.com/hashicorp/waypoint-plugin-sdk/docs"
+)
+
+// TaskLauncher implements the TaskLauncher plugin interface to support
+// launching on-demand tasks for the Waypoint server.
+type TaskLauncher struct {
+	config TaskLauncherConfig
+}
+
+// StartTaskFunc implements component.TaskLauncher
+func (p *TaskLauncher) StartTaskFunc() interface{} {
+	return p.StartTask
+}
+
+// StopTaskFunc implements component.TaskLauncher
+func (p *TaskLauncher) StopTaskFunc() interface{} {
+	return p.StopTask
+}
+
+// TaskLauncherConfig is the configuration structure for the task plugin. At
+// this time all these are simply copied from what the Waypoint Server
+// installation is using, with the only exception being the OdrTaskRoleName.
+type TaskLauncherConfig struct {
+	// Cluster is the ECS we're operating in
+	Cluster string `hcl:"cluster,optional"`
+
+	// Region is the AWS region we're operating in, e.g. us-west-2, us-east-1
+	Region string `hcl:"region,optional"`
+
+	// ExecutionRoleName is the name of the AWS IAM role to apply to the task's
+	// Execution Role. At this time we reuse the same Role as the Server
+	// Execution Role.
+	ExecutionRoleName string `hcl:"execution_role_name,optional"`
+
+	// TaskRoleName is the name of the AWS IAM role to apply to the task. This
+	// role determines the privileges the ODR builder has, and must have the correct
+	// policies in place to work with the provided registries.
+	TaskRoleName string `hcl:"task_role_name,optional"`
+
+	// Subnets are the list of subnets for the cluster. These will match the
+	// subnets used for the Cluster
+	Subnets string `hcl:"subnets,optional"`
+
+	// SecurityGroupId is the security group used for the Waypoint tasks.
+	SecurityGroupId string `hcl:"security_group_id,optional"`
+
+	// LogGroup is the CloudWatch log group name to use.
+	LogGroup string `hcl:"log_group,optional"`
+}
+
+func (p *TaskLauncher) Documentation() (*docs.Documentation, error) {
+	doc, err := docs.New(
+		docs.FromConfig(&TaskLauncherConfig{}),
+		docs.FromFunc(p.StartTaskFunc()),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	doc.Description(`
+Launch an ECS task for on-demand tasks from the Waypoint server.
+
+This will use the standard AWS environment variables and IAM Role information to
+source authentication information for AWS, using the configured task role.
+If no task role name is specified, Waypoint will create one with the required
+permissions.
+`)
+
+	doc.SetField(
+		"odr_task_role_name",
```

Comment:
Given this is just the generic task plugin, I think let's just change the name of this to `task_role_name` since in the future, it might be used for things other than odr.

Reply:
Oh, Also, I see that the field is actually called task_role_name in the config too, so the docs are just out of date.

Reply:
Yep good catch, cleaned up in 45509b65